detox/src/Detox.js

```
@@ -32,75 +32,44 @@ class Detox {
 
     this.device = null;
   }
 
-  async init(userParams) {
-    const sessionConfig = await this._getSessionConfig();
-    const params = {
-      launchApp: true,
-      initGlobals: true,
-      ...userParams,
-    };
-
-    if (!this._userSession) {
-      this._server = new DetoxServer({
-        log: logger,
-        port: new URL(sessionConfig.server).port,
-      });
+  init(userParams) {
+    if (!this._initPromise) {
+      this._initPromise = Promise.race([
+        this._doInit(userParams),
+        new Promise((_, reject) => {
+          this._rejectInit = reject;
+        }),
+      ]);
     }
 
-    this._client = new Client(sessionConfig);
-    this._client.setNonresponsivenessListener(this._onNonresnponsivenessEvent.bind(this));
-    await this._client.connect();
-
-    const DeviceDriverClass = DEVICE_CLASSES[this._deviceConfig.type];
-    if (!DeviceDriverClass) {
-      throw new Error(`'${this._deviceConfig.type}' is not supported`);
-    }
-
-    const deviceDriver = new DeviceDriverClass({
-      client: this._client,
-    });
-
-    this._artifactsManager.subscribeToDeviceEvents(deviceDriver);
-    this._artifactsManager.registerArtifactPlugins(deviceDriver.declareArtifactPlugins());
-
-    const device = new Device({
-      deviceConfig: this._deviceConfig,
-      deviceDriver,
-      sessionConfig,
-    });
-
-    await device.prepare(params);
-
-    const globalsToExport = {
-      ...deviceDriver.matchers,
-      device,
-    };
-
-    Object.assign(this, globalsToExport);
-    if (params.initGlobals) {
-      Object.assign(global, globalsToExport);
-    }
-
-    await this._artifactsManager.onInit();
-
-    return this;
+    return this._initPromise;
   }
 
   async cleanup() {
-    await this._artifactsManager.onBeforeCleanup();
+    await this._assertNoPendingInit().catch(_.noop);
+
+    if (this._artifactsManager) {
+      await this._artifactsManager.onBeforeCleanup();
+      this._artifactsManager = null;
```

Nullifying now everything, to fail fast if somebody tries to manipulate a particular Detox instance after the cleanup.